### ISSUE TITLE
Expand JSON e2e tests to cover all control plane components

### ISSUE THREAD

**Issue by serathius** [kubernetes/kubernetes#100748]
With recent migration of Kubelet to structured logging https://github.com/kubernetes/kubernetes/issues/99802 introduced e2e tests that enable JSON in Kubelet.

Organized migration of other components haven't yet started and is on best effort bases, yet some line were already migrated. 
As code-base will be slowly migrated, we need to expand the test to avoid errors like https://github.com/kubernetes/kubernetes/pull/100739

I propose to expand existing e2e test suite ([kind-json-logging-master](https://testgrid.k8s.io/sig-instrumentation-tests#kind-json-logging-master) and [kind-json-logging-1.21](https://testgrid.k8s.io/sig-instrumentation-tests#kind-json-logging-1.21)) to cover all control plane components (apiserver, controller-manager and scheduler).

/cc @BenTheElder

**Comment by BenTheElder** [comment]
If all components have this flag do we just want to enable it on all components?
We already increase the log level on all components in CI, it should be pretty straightforward to expand this to all:
- pick a new environment variable name for the script with a more appropriate name for all components instead of just kubelet
- set it on the job in addition to the old kubelet one, by sending a PR to test-infra
- send a PR to the kind script to set the value in all places where we set verbosity instead of just kubelet, switching to the new variable name

**Comment by sladyn98** [comment]
@serathius So currently the e2e tests do not actually read and test what the logs return the current job added just makes sure that kubelet is running https://github.com/kubernetes/test-infra/pull/21332/files#r594444594. Would we want to write similar tests for the different control planes while we start migration or do you have something else in mind with respect to the automated testing ?

**Comment by serathius** [comment]
Yes, currently we just test if Kubelet is running to test that flag is working and if enabling json doesn't cause panics. This exact issue proposes to expand this to other control plane, but your idea would be a great follow up. Would be great to propose a test plan that will specify what we want to test and how. If you have ideas some ideas could you collect hem in some issue so we can discuss.

**Comment by BenTheElder** [comment]
There's no portable way to get these logs. There are non portable ways but those cannot be used in a standard e2e test.

**Comment by sladyn98** [comment]
Oh so for the standard e2e tests it might be good enough to just test the startups of these control planes with the logging enabled.

**Comment by sladyn98** [comment]
@serathius I was thinking that we could get started with maybe writing similar to https://github.com/kubernetes/test-infra/pull/21332 one for the remaining components and then maybe find a way of checking the logs generated as we have bigger discussions
a) kube-apiserver --> Completed
b) etcd --- > TODO
c) kube-scheduler ---> TODO
d) kube-controller-manager --> TODO

**Comment by serathius** [comment]
sound good. 
FYI Etcd is not a kubernetes component as so it's out of scope of this issue.

**Comment by serathius** [comment]
/wg structured-logging

@BenTheElder As getting core components logs is not portable, we could test a container with same logging library and using Kubenetes logs API to access logs. Created https://github.com/kubernetes/kubernetes/issues/102237 to track it

**Comment by BenTheElder** [comment]
it still might not be a terrible idea to write a simple test script to probe the exported logs from a cluster with json logging enabled for any non-json lines once we know that everything should be. I think this could probably be hacked up with jq or something, but I wouldn't position it like a normal e2e test.

**Comment by serathius** [comment]
Could you give a hint how to implement and run such script?

**Comment by serathius** [comment]
/priority important-soon
This issue is based on feedback from Kubelet migration retrospection, it is required before organizing migration of next component.
/cc @kubernetes/wg-structured-logging-members

**Comment by sladyn98** [comment]
> it still might not be a terrible idea to write a simple test script to probe the exported logs from a cluster with json logging enabled for any non-json lines once we know that everything should be. I think this could probably be hacked up with jq or something, but I wouldn't position it like a normal e2e test.

Would this follow the same flow as in `logs/json/json_test.go`  where we  check if json logging is enabled and then run jq separately or within the same flow. 
To be more specific: 

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/logs/json/json_test.go#L36

Seems to be getting the data and validating it so in that case would we be running `jq` after grabbing the logs ?

**Comment by serathius** [comment]
For now let's focus on the original tasks of having ([kind-json-logging-master](https://testgrid.k8s.io/sig-instrumentation-tests#kind-json-logging-master) and [kind-json-logging-1.21](https://testgrid.k8s.io/sig-instrumentation-tests#kind-json-logging-1.21)) test scenarios, also configure other components.

**Comment by rahulii** [comment]
Would like to take this one!
/cc @serathius

**Comment by rahulii** [comment]
@serathius just created a PR: https://github.com/kubernetes-sigs/kind/pull/2335

**Comment by rahulii** [comment]
Components to be picked:
- [ ] kube-scheduler
- [ ] kube-apiserver
- [ ] controller-manager

**Comment by pchan** [comment]
Attempted to cover the remaining components in this [PR 2387](https://github.com/kubernetes-sigs/kind/pull/2387)

**Comment by serathius** [comment]
It would be great if we had e2e setup for all components before we start structured migration next week. @pchan @rahulii will you be able to work on it this week? Please let me know if there is anything I can help with.

**Comment by serathius** [comment]
@pchan @rahulii  Thanks for help with control plane components, for 1.23 we are planning to migrate kube-proxy to structured logging. Could you also try to enable JSON logging on kube-proxy?

**Comment by serathius** [comment]
Confirmed that in last CI run https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-json-logging/1435884240037220352

Logs from kubelet and control plane components are written in JSON:
* [kubelet](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-json-logging/1435884240037220352/artifacts/logs/kind-control-plane/kubelet.log)
* [scheduler](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-json-logging/1435884240037220352/artifacts/logs/kind-control-plane/containers/kube-scheduler-kind-control-plane_kube-system_kube-scheduler-486851ae4eab5da94d5979956cdc6248e73311d003ae1d80024df02c2e1fe2db.log)
* [controller-manager](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-json-logging/1435884240037220352/artifacts/logs/kind-control-plane/containers/kube-controller-manager-kind-control-plane_kube-system_kube-controller-manager-b48404f4b6a9a35914a64e8b8478ed80760ae71190798d926e7ef03c85ded1b5.log)
* [apiserver](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-kind-e2e-json-logging/1435884240037220352/artifacts/logs/kind-control-plane/containers/kube-apiserver-kind-control-plane_kube-system_kube-apiserver-e00e1f17b55aeb007af0ab84520f8274aa0f198b3a0dbd6e35de5f5e6d1a980d.log)

Great Job @pchan @rahulii!
Could you take a look into changing log format of kube-proxy?

**Comment by pchan** [comment]
>
> Could you take a look into changing log format of kube-proxy?
>

I will work on getting JSON logging for kube-proxy.

>

**Comment by pchan** [comment]
Filed #105165 for kube-proxy. Are we good to close this issue ?

**Comment by serathius** [comment]
sg
